neuromation/cli/utils.py

```diff
@@ -680,3 +683,164 @@ def pager_maybe(
+        (Select Graphic Rendition) ones.
+    """
+
+    style_re = re.compile(r"(\033\[(?:\d|;)*m)")
+    ansi_reset_all = "\033[0m"
+    whitespace = "\t\n\x0b\x0c\r "
+
+    @classmethod
+    def is_styled(cls, text: str) -> bool:
+        return "\033" in text
+
+    # As we often do operations on headers or similar static text it makes sense to
+    # cache
+    @classmethod
+    @lru_cache(255)
+    def unstyle(cls, text: str) -> str:
+        return click.unstyle(text)
+
+    @classmethod
+    @lru_cache(255)
+    def width(cls, text: str) -> int:
+        if cls.is_styled(text):
+            return wcswidth(cls.unstyle(text))
+        else:
+            return wcswidth(text)
+
+    @classmethod
+    def ljust(cls, text: str, width: int) -> str:
+        if cls.is_styled(text):
+            return text.ljust(width + len(text) - len(cls.unstyle(text)))
+        else:
+            return text.ljust(width)
+
+    @classmethod
+    def rjust(cls, text: str, width: int) -> str:
+        if cls.is_styled(text):
+            return text.rjust(width + len(text) - len(cls.unstyle(text)))
+        else:
+            return text.rjust(width)
+
+    @classmethod
+    def center(cls, text: str, width: int) -> str:
+        if cls.is_styled(text):
+            return text.center(width + len(text) - len(cls.unstyle(text)))
+        else:
+            return text.center(width)
+
+    @classmethod
+    def trim(cls, text: str, width: int) -> str:
+        if not cls.is_styled(text):
+            return text[:width]
+
+        result = []
+        has_unclosed = False
+        remaining = width
+        for token in cls.style_re.split(text):
+            if cls.style_re.match(token):
+                result.append(token)
+                # We will need to add a reset at the end if there are any non-closed
+                # sequences
+                has_unclosed = token != cls.ansi_reset_all
+            else:
+                if len(token) >= remaining:
+                    result.append(token[:remaining])
+                    remaining = 0
+                    break
+                remaining -= len(token)
+                result.append(token)
+
+        if has_unclosed:
+            result.append(cls.ansi_reset_all)
+
+        return "".join(result)
+
+    @classmethod
+    def wrap(cls, text: str, width: int) -> Sequence[str]:
+        if cls.is_styled(text):
+            # Fast return if we don't need to wrap the text
+            if cls.width(text) <= width:
+                return [text]
+
+            wrapped = fill(cls.unstyle(text), width)
+            return list(cls._reapply_styles(text, wrapped))
+        else:
+            return wrap(text, width)
+
+    @classmethod
+    def _reapply_styles(cls, text: str, wrapped: str) -> Iterator[str]:
+        """ Iterate over wrapped text and original one and reapply styles from
+        the original to wrapped
+        """
+        sm = SequenceMatcher(None, text, wrapped)
```

Comment:
Found using difflib to be way more readable than trying to write a simpler one myself.